lahja/asyncio/endpoint.py

```diff
@@ -313,19 +317,23 @@ def run(self, *args, **kwargs):  # type: ignore
         await self._receiving_loop_running.wait()
         self.logger.debug("Endpoint[%s]: running", self.name)
 
-    def _stop(self) -> None:
+    async def _stop(self) -> None:
         """
         Stop the :class:`~lahja.endpoint.asyncio.AsyncioEndpoint` from receiving further events.
         """
         if not self.is_running:
             return
 
-        self._stop_server()
+        await self._stop_server()
 
         self._running = False
 
-        for task in self._endpoint_tasks:
+        for task in tuple(self._endpoint_tasks):
```

Comment:
Curious why we need to transform the `set` into a `tuple` here?

Reply:
because by doing the cancellation properly the task actually finishes, executing it's cleanup task which removes it from the list of tasks and then python complains that the size changed during iteration so we have to freeze it before iterating over it.